paasta_tools/kubernetes/application/controller_wrappers.py

```diff
@@ -196,9 +198,19 @@ def create(self, kube_client: KubeClient) -> None:
         self.ensure_pod_disruption_budget(kube_client)
         self.sync_horizontal_pod_autoscaler(kube_client)
 
+    def deep_delete_and_wait(self, kube_client: KubeClient) -> None:
+        self.deep_delete(kube_client)
+        while self.kube_deployment in set(list_all_deployments(kube_client)):
+            self.logging.debug(
+                "waiting for instance {self.kube_deployment.service} {self.kube_deployment.instance} to die."
+            )
+            sleep(0.5)
+
     def update(self, kube_client: KubeClient) -> None:
         # If autoscaling is enabled, do not update replicas.
         # In all other cases, replica is set to max(instances, min_instances)
+        if self.get_soa_config().get("bounce_method", "") == "brutal":
```

Comment:
Please add a thread here. deep_delete_and_wait shouldn't block other things. And don't forget to add tests for this.